### PR TITLE
refactor!: add new resolver for IDEs

### PR DESCRIPTION
**Pull request by zakuciael**
## TL;DR
This pull request changes the way the plugin resolves information about installed IDEs and recently opened projects. To do so it now utilizes two files available for every JetBrains IDE which are `product-info.json` and `idea.properties`.

## What changed
- **[BREAKING CHANGE]** Removed `jetbrains-shell-scripts-path`, `jetbrains-android-studio-config-path`, and `jetbrains-configs-path` config options in favor of `jetbrains-install-dir`.
- Moved `MapToErrorLog` trait to `src/traits.rs` file to avoid name conflicts with the `error!()` macro.
- Expanded `ide` module to sub-modules:
  - `data` - Implementing `IDEData` struct holding all the resolved information about an IDE.
  - `de` - Implementing the `Deserialize` trait for `IDEType` and `evaluate_property()` method for deserializing Java properties containing expressions.
  - `product_info` - Implementing `IDEProductInfo` struct defining fields of the `product-info.json` file.
  - `properties` - Implementing `IDEProperties` struct defining fields of the `idea.properties` file.
  - The root module - Implementing the `IDEType` enum
- The `RecentProject` struct now holds an `Arc` to the IDE from which it was resolved.
- Implemented a way to display fallback IDE icons.
- Add caching for `icon_name`

## Why make this change
This change makes the plugin **almost autonomous** in terms of new IDE releases, only requiring an update to the `IDEType` enum to add new **product code**. It also lowers the amount of guesswork and user configuration required to resolve information needed for the plugin to work.

### CODE REVIEW

**Comment by zakuciael** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3** <a href="https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/3" target="_blank">Graphite</a>  👈
    * **PR #4** <a href="https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/4" target="_blank">Graphite</a> 
      * **PR #5** <a href="https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/5" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/3?utm_source=stack-comment).

**Comment by zakuciael** [comment]
### Merge activity

* **Oct 11, 12:10 PM**: @zakuciael started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/3).
* **Oct 11, 12:10 PM**: [Graphite](https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/3) couldn't merge this PR because it was not satisfying all requirements.
* **Oct 11, 12:12 PM**: @zakuciael started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/3).
* **Oct 11, 12:12 PM**: [Graphite](https://app.graphite.dev/github/pr/zakuciael/rofi-jetbrains/3) couldn't merge this PR because it was not satisfying all requirements.
